tools/cargo-to-recipe.sh

```diff
@@ -0,0 +1,287 @@
+#!/usr/bin/env bash
+
+die() {
+	printf '%s' "${*+$'\e[31mError: \e[0m'$*$'\n'}"
+	(( usage )) && usage
+	exit 1
+} 1>&2
+
+usage() {
+	cat <<- EOF
+	Usage: $0 [options] URI category/port
+
+	Creates a recipe template for a crates.io package, filled with
+	information at hand.
+
+	Options:
+	  -h, --help	show this help message and exit
+	  -k, --keep	keep the generated temporary directory
+	  -psd, --print-source-directories
+	 		also print SOURCE_DIRs
+	  -c CMD, --cmd=CMD
+	 		specify the command runtime
+	  -b PORTNAME, --bump PORTNAME
+	 		bump the crates.io dependencies of the specified port
+	EOF
+}
+
+temp() { rm -rf "$tempdir"; }
+
+. "$(finddir B_USER_SETTINGS_DIRECTORY)"/haikuports.conf
+while (( $# )); do
+	case "$1" in
+		-h|--help)
+			usage
+			exit 0
+			;;
+		-k|--keep)
+			temp() { printf '%s\n' "Kept $tempdir"; }
+			;;
+		-psd|--print-source-directories)
+			psd=3
+			;;
+		-c)
+			shift
+			;&
+		--cmd=*)
+			cmd=${1#*=}
+			;;
+		-b|--bump)
+			portName=$2
+			directory=$(
+				find "$TREE_PATH" -mindepth 3 -maxdepth 3 \
+					-iname "$portName-*.*.recipe" |
+					awk 'FNR == 1 { gsub("/[^/]*$", "")
+							print }'
+			)
+			bump=1
+			shift
+			;;
+		*://*)
+			SOURCE_URI=$1
+			;;
+		*-*/*)
+			directory=$TREE_PATH/$1
+			: "${1//-/_}"
+			portName=${_#*/}
+			;;
+		-?*)
+			usage=1 die "Invalid option."
+			;;
+		*)
+			usage=1 die "Invalid argument."
+			;;
+	esac
+	shift
+	false
+done && usage=1 die
+
+mkdir -p "$directory"/download
+cd "$directory" || die "Invalid port directory."
+
+if (( bump )); then
+	eval "recipe=$(
+		ls -v --quoting-style=shell-escape \
+			"$portName"-[0-9]*.[0-9]*.recipe | tail -n 1
```

Comment:
Suggestion:
```diff
- 			"$portName"-[0-9]*.[0-9]*.recipe | tail -n 1
+ 			"$portName"-*.recipe | tail -n 1
```
(This would allow more special characters, e.g. `_` and `~`, instead of just digits.)

BTW, please forgive me for not doing a full review.

Reply:
Well, the glob does account for other characters; just not at the start of it (e.g., `0.5.0~git` will be matched). My problem with this is that I'm not sure how to get the most recent version (how does `haikuporter` do it?), and I think that Rust packages are supposed to have the `x.x.x` versioning schemes, anyway.

Reply:
Please do not reimplement parsing logic. HaikuPorter has an extensive internal library of queries, many of which are exposed to the command line. If you are missing something, please make a feature request against HaikuPorter so it can be implemented centrally. The same goes, even more so, for dependency logic.

Reply:
@mmlr The whole point of this tool was to parse `Cargo.toml` to get the dependency info then format it to `SOURCE_URI`s and `CHECKSUM_SHA256`s, though. Are you saying that this should be implemented in HaikuPorter as well?